In [1]:
import os
# import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from ultralytics import YOLO
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report

In [2]:
data_yaml_path = "data.yaml"
model_name = "yolov8n.pt"
epochs = 1
image_size = 128

model = YOLO(model_name)
model.train(data=data_yaml_path, epochs=epochs, imgsz=image_size)

New https://pypi.org/project/ultralytics/8.3.50 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.48 🚀 Python-3.8.10 torch-2.4.1+cu121 CPU (Intel Core(TM) i7-3537U 2.00GHz)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=data.yaml, epochs=1, time=None, patience=100, batch=16, imgsz=128, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train5, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, s

[W1217 15:20:58.517533230 NNPACK.cpp:61] Could not initialize NNPACK! Reason: Unsupported hardware.


Model summary: 225 layers, 3,011,238 parameters, 3,011,222 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir runs/detect/train5', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'


train: Scanning /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/train/labels.cache... 23064 images, 11221 backgrounds, 0 corrupt: 100%|██████████| 23064/23064 [00:00<?, ?it/s]
/home/lex/.local/lib/python3.8/site-packages/torch/cuda/__init__.py:654: UserWarning: Can't initialize NVML
  warnings.warn("Can't initialize NVML")
val: Scanning /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/valid/labels.cache... 5948 images, 2866 backgrounds, 0 corrupt: 100%|██████████| 5948/5948 [00:00<?, ?it/s]


Plotting labels to runs/detect/train5/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 128 train, 128 val
Using 0 dataloader workers
Logging results to runs/detect/train5
Starting training for 1 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/1         0G      2.432      2.531       1.32         10        128: 100%|██████████| 1442/1442 [33:23<00:00,  1.39s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 186/186 [03:51<00:00,  1.24s/it]


                   all       5948       4583      0.152       0.16     0.0865     0.0247

1 epochs completed in 0.624 hours.
Optimizer stripped from runs/detect/train5/weights/last.pt, 6.2MB
Optimizer stripped from runs/detect/train5/weights/best.pt, 6.2MB

Validating runs/detect/train5/weights/best.pt...
Ultralytics 8.3.48 🚀 Python-3.8.10 torch-2.4.1+cu121 CPU (Intel Core(TM) i7-3537U 2.00GHz)
Model summary (fused): 168 layers, 3,006,038 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 186/186 [03:45<00:00,  1.21s/it]


                   all       5948       4583      0.152       0.16     0.0864     0.0247
                 Smoke         88        145     0.0172     0.0483    0.00564    0.00252
              Wildfire       3005       4438      0.286      0.272      0.167     0.0468
Speed: 0.3ms preprocess, 30.3ms inference, 0.0ms loss, 0.7ms postprocess per image
Results saved to runs/detect/train5


ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7fa2a546cbe0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.04804

In [6]:
# Загрузка наилучшей версии модели для оценки характеристик
yolo_model = YOLO('./models/yolo-8epochs-image128px.pt')

test_image_dir = './datasets/test/images'
predicts_for_files = []
y_pred = []
y_test = []

def get_basefilename(filename):
    if filename.lower().endswith('.jpeg'):
        return filename[:-5]
    return filename[:-4]

# Нужно преобразовать массив вероятностей Y 
# (где каждая вероятность показывает уверенность в наличии пожара в конкретной области)
# в одно число, которое будет отражать общую уверенность в том, что “где-то что-то есть” на изображении.
# Используем Вероятность объединения событий (ИЛИ)
def calculate_combined_probability(confidences):
    if not confidences:
        # Если нет объектов, вероятность 0
        return 0

    # Вероятность отсутствия пожара в каждой области: 1 - confidence
    no_fire_probabilities = [1 - conf for conf in confidences]

    # Вероятность того, что нет пожара нигде:
    combined_no_fire_prob = np.prod(no_fire_probabilities)

    # Вероятность того, что хотя бы где-то есть пожар:
    combined_probability = 1 - combined_no_fire_prob
    return combined_probability

# i = 0
for filename in os.listdir(test_image_dir):
    if filename.endswith(('.jpg', '.jpeg', '.png')):  # Обрабатываем только изображения
        image_path = os.path.join(test_image_dir, filename)
        try:
            results = yolo_model.predict(source=image_path)
            confidences = []
            for result in results:
                boxes = result.boxes
                if boxes:
                    confidences.extend([box.conf.item() for box in boxes])
            predicts_for_files.append({
                'basefilename': get_basefilename(filename),
                'pred': calculate_combined_probability(confidences),
                'confidences': confidences,
                'test': None
            })
        except Exception as e:
            print(f"Ошибка при обработке изображения {filename}: {e}")
        # i += 1
        # if i >= 200:
        #     break

test_labels_dir = './datasets/test/labels'
for predicts in predicts_for_files:
    filename = predicts['basefilename'] + '.txt'
    try:
        with open(os.path.join(test_labels_dir, filename), "r") as f:
            lines = f.readlines()
            if lines:
                data = lines[0].strip().split()
                confidence = 1 if len(data) > 0 else 0
                y_test.append(confidence)
                predicts['test'] = confidence
            else:
                predicts['test'] = 0
    except Exception as e:
        print(f"Ошибка при чтении файла {filename}: {e}")

# ----------------------------------------------------------------------------------
# Интерпретируем результаты модели таким образом, что
# если она указывает на какую-то область, значит она уже утверждает что что-то нашла, т.е. pred > 0.5
# Это важно, потому что вероятность с которой модель даёт ответ включает в себя не только уверенность в наличии пожара
# но и уверенность в месте его расположения.
# Такая "уверенность" модели всегда будет заведомо меньше вероятности просто наличия пожара
for x in predicts_for_files:
    x['pred0'] = (1 + x['pred']) / 2 if x['pred'] > 0 else 0
# ----------------------------------------------------------------------------------

y_pred = [1 if x['pred0'] >= 0.5 else 0 for x in predicts_for_files]
y_test = [x['test'] for x in predicts_for_files]
y_pred, y_test = np.array(y_pred), np.array(y_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='binary')

print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")

report = classification_report(y_test, y_pred, target_names=["Class 0", "Class 1"])

print("\nClassification Report:\n")
print(report)


image 1/1 /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/test/images/snapshot-2016-07-30_jpg.rf.df158f3fee405f1e5ce68555cfbf24e0.jpg: 128x128 (no detections), 234.7ms
Speed: 1.2ms preprocess, 234.7ms inference, 0.7ms postprocess per image at shape (1, 3, 128, 128)

image 1/1 /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/test/images/cal_north2021_240_7_png_jpg.rf.072cedafbe5d20272a3554f55fa427d1.jpg: 128x128 2 Wildfires, 50.5ms
Speed: 4.0ms preprocess, 50.5ms inference, 1.1ms postprocess per image at shape (1, 3, 128, 128)

image 1/1 /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/test/images/ETN_snapshot-2005-07-25_jpg.rf.beb6fe6837c93a4ecc3b30068b5c6304.jpg: 128x128 1 Wildfire, 253.9ms
Speed: 7.1ms preprocess, 253.9ms inference, 1.4ms postprocess per image at shape (1, 3, 128, 128)

image 1/1 /home/lex/Рабочий стол/work/Магистратура/Хакатон/t3/datasets/test/images/2021_243_23_png_jpg.rf.1b20cf47614e0dc6c2baf333c69a4838.jpg: 128x128 (no detect

## Оценка модели:

- Accuracy: 0.8017711171662125
- Precision: 0.9285714285714286
- Recall: 0.6679841897233202
- F1 Score: 0.7770114942528736

Classification Report:

              precision    recall  f1-score   support

     Class 0       0.73      0.94      0.82       709
     Class 1       0.93      0.67      0.78       759

    accuracy                           0.80      1468
    
    macro avg      0.83      0.81      0.80      1468
    weighted avg   0.83      0.80      0.80      1468